File: src/nl/rubensten/texifyidea/structure/LatexPresentationFactory.java
package nl.rubensten.texifyidea.structure;

import com.intellij.navigation.ItemPresentation;
import nl.rubensten.texifyidea.psi.LatexCommands;

/**
 * @author Ruben Schellekens
 */
public class LatexPresentationFactory {

    public static ItemPresentation getPresentation(LatexCommands commands) {
        switch (commands.getCommandToken().getText()) {
            case "\\section":
                return new LatexSectionPresentation(commands);
            case "\\subsection":
                return new LatexSubSectionPresentation(commands);
            case "\\subsubsection":
                return new LatexSubSubSectionPresentation(commands);
            case "\\paragraph":
                return new LatexParagraphPresentation(commands);
            case "\\subparagraph":
                return new LatexSubParagraphPresentation(commands);
            case "\\newcommand":
                return new LatexNewCommandPresentation(commands);
            case "\\label":
                return new LatexLabelPresentation(commands);
            case "\\include":
                return new LatexIncludePresentation(commands);
            case "\\includeonly":
                return new LatexIncludePresentation(commands);
            case "\\input":
                return new LatexIncludePresentation(commands);
        }

        throw new IllegalArgumentException(
                "Commands " + commands.getCommandToken().getText() + "" + " not supported");
    }

    private LatexPresentationFactory() {
        throw new AssertionError("Nope");
    }

}

File: src/nl/hannahsten/texifyidea/structure/latex/LatexPresentationFactory.kt
package nl.hannahsten.texifyidea.structure.latex

import com.intellij.navigation.ItemPresentation
import nl.hannahsten.texifyidea.TexifyIcons
import nl.hannahsten.texifyidea.lang.predefined.CommandNames
import nl.hannahsten.texifyidea.psi.LatexCommands
import nl.hannahsten.texifyidea.util.labels.getLabelPosition
import nl.hannahsten.texifyidea.util.magic.CommandMagic

/**
 * @author Hannah Schellekens
 */
object LatexPresentationFactory {

    @JvmStatic
    fun getPresentation(commands: LatexCommands): ItemPresentation {
        // Any delay here will be a delay before the contents of the structure popup are shown

        val labelPosition = commands.getLabelPosition()
        if (labelPosition >= 0) {
            return LatexLabelPresentation(commands, labelPosition)
        }
        return CommandNames.run {
            when (commands.name) {
                PART -> LatexPartPresentation(commands)
                CHAPTER -> LatexChapterPresentation(commands)
                SECTION -> LatexSectionPresentation(commands)
                SUB_SECTION -> LatexSubSectionPresentation(commands)
                SUB_SUB_SECTION -> LatexSubSubSectionPresentation(commands)
                PARAGRAPH -> LatexParagraphPresentation(commands)
                SUB_PARAGRAPH -> LatexSubParagraphPresentation(commands)
                NEW_COMMAND, RENEW_COMMAND, DECLARE_MATH_OPERATOR, NEW_DOCUMENT_COMMAND -> LatexNewCommandPresentation(commands)
                DECLARE_PAIRED_DELIMITER, DECLARE_PAIRED_DELIMITER_X, DECLARE_PAIRED_DELIMITER_XPP -> LatexPairedDelimiterPresentation(commands)
                LABEL -> LatexLabelPresentation(commands, labelPosition)
                BIB_ITEM -> BibitemPresentation(commands)
                in CommandMagic.allFileIncludeCommands -> LatexIncludePresentation(commands)
                else -> LatexOtherCommandPresentation(commands, TexifyIcons.DOT_COMMAND)
            }
        }
    }
}